### ISSUE TITLE
Document (or fix) stack safety issues

### ISSUE THREAD

**Issue by milesfrain** [purescript/purescript-lists#194]
Some example of existing stack-unsafe functions:
- `sort` - reported in #192 
- `nubEq`, can be fixed for strict lists by applying same strategy used for struct `nub` in #179 
- `nub` for lazy lists. Don't know of a good way to fix this. The strict `nub` strategy of #179 cannot be applied here because you can't `reverse` an infinite lazy list.

We should ideally setup stack-safety tests for every function. To help make those tests run faster, we could shrink the v8 stack size with something like `--stack_size=10` (default is 984 kB).

**Comment by hdgarrood** [comment]
I would prefer that we test with lists which are large enough to blow the stack than shrink the stack artificially, as we ideally ought to be testing with large inputs anyway, e.g. so that we are more likely to notice if we make a change which unintentionally make some functions much slower for large inputs.

**Comment by imcotton** [comment]
From **Data.List.Lazy**

- `elemIndex`
- `elemLastIndex`
- `findLastIndex`
- `findIndex`  <---  caused by

https://github.com/purescript/purescript-lists/blob/6383c4f202b3f69474f9f7da182c2d42fcc3111c/src/Data/List/Lazy.purs#L317-L327

**Comment by liftM** [comment]
`group` also does not appear to be stack-safe.

**Comment by liftM** [comment]
I might be crazy, but... it appears that `concat` is also not stack-safe?

**Comment by garyb** [comment]
Yeah, seems so - `concat` is implemented with `bind` which isn't stack safe: https://github.com/purescript/purescript-lists/blob/68bce83550247f1bcaf2c160b01e6740f1f5d471/src/Data/List/Types.purs#L165-L167
